mlir/lib/Catalyst/Driver/CompilerDriver.cpp

```diff
@@ -0,0 +1,336 @@
+
+std::string joinPasses(const Pipeline::PassList &passes)
+{
+    std::string joined;
+    llvm::raw_string_ostream stream{joined};
+    llvm::interleaveComma(passes, stream);
+    return joined;
+}
+
+struct CatalystIRPrinterConfig : public PassManager::IRPrinterConfig {
+    typedef std::function<LogicalResult(Pass *, PrintCallbackFn print)> PrintHandler;
+    PrintHandler printHandler;
+
+    CatalystIRPrinterConfig(PrintHandler printHandler)
+        : IRPrinterConfig(/*printModuleScope=*/true), printHandler(printHandler)
+    {
+    }
+
+    void printAfterIfEnabled(Pass *pass, Operation *operation, PrintCallbackFn printCallback) final
+    {
+        if (failed(printHandler(pass, printCallback))) {
+            operation->emitError("IR printing failed");
+        }
+    }
+};
+
+} // namespace
+
+namespace {
+/// Parse an MLIR module given in textual ASM representation. Any errors during parsing will be
+/// output to diagnosticStream.
+OwningOpRef<ModuleOp> parseMLIRSource(MLIRContext *ctx, StringRef source, StringRef moduleName,
+                                      raw_ostream &diagnosticStream)
+{
+    auto moduleBuffer = llvm::MemoryBuffer::getMemBufferCopy(source, moduleName);
+    auto sourceMgr = std::make_shared<llvm::SourceMgr>();
+    sourceMgr->AddNewSourceBuffer(std::move(moduleBuffer), SMLoc());
+
+    FallbackAsmResourceMap fallbackResourceMap;
+    ParserConfig parserConfig{ctx, /*verifyAfterParse=*/true, &fallbackResourceMap};
+
+    SourceMgrDiagnosticHandler sourceMgrHandler(*sourceMgr, ctx, diagnosticStream);
+    return parseSourceFile<ModuleOp>(sourceMgr, parserConfig);
+}
+
+/// Parse an LLVM module given in textual representation. Any parse errors will be output to
+/// the provided SMDiagnostic.
+std::unique_ptr<llvm::Module> parseLLVMSource(llvm::LLVMContext &context, StringRef source,
+                                              StringRef moduleName, llvm::SMDiagnostic &err)
+{
+    auto moduleBuffer = llvm::MemoryBuffer::getMemBufferCopy(source, moduleName);
+    return llvm::parseIR(llvm::MemoryBufferRef(*moduleBuffer), err, context);
+}
+
+/// Register all dialects required by the Catalyst compiler.
+void registerAllCatalystDialects(DialectRegistry &registry)
+{
+    // MLIR Core dialects
+    registerAllDialects(registry);
+    registerAllExtensions(registry);
+
+    // HLO
+    mhlo::registerAllMhloDialects(registry);
+    stablehlo::registerAllDialects(registry);
+
+    // Catalyst
+    registry.insert<CatalystDialect>();
+    registry.insert<quantum::QuantumDialect>();
+    registry.insert<gradient::GradientDialect>();
+}
+} // namespace
+
+FailureOr<llvm::Function *> getJITFunction(MLIRContext *ctx, llvm::Module &llvmModule)
+{
+    Location loc = NameLoc::get(StringAttr::get(ctx, llvmModule.getName()));
+    for (auto &function : llvmModule.functions()) {
+        emitRemark(loc) << "Found LLVM function: " << function.getName() << "\n";
+        if (function.getName().starts_with("jit_")) {
```

Comment:
I think this would break it. I know that isn't great, but we need some way of determining which function is the entry point when compiling from LLVM IR. I'm definitely open to suggestions 🙂

Reply:
We can annotate the entry point from the frontend and look for the annotation here.

Reply:
After some checks I think that we probably need to allow users to specify a signature of a top-level function. What do you think?

Some details: I see an assertion failure [here](https://github.com/PennyLaneAI/catalyst/blob/1abfbc69c484ba8f4c6310b2eaa3e36ed0ec9748/frontend/catalyst/compilation_pipelines.py#L656). The failing test checks [compilation from IR string](https://github.com/PennyLaneAI/catalyst/blob/1abfbc69c484ba8f4c6310b2eaa3e36ed0ec9748/frontend/test/pytest/test_compiler.py#L297).  The shape there refers to a Pytree shape, and user should already know it at the time they write IR string.

Btw, do you think that the name `shape` better associates with tensor shapes rather than with the Pytree shapes? How about renaming it to `pytree_shape` or something like that? CC @maliasadi

Reply:
Regarding the first comments, I'm curious about the entry point. I would think the compiler driver's job ends at producing the shared library. Are you saying we would also like to be able to use the frontend machinery to load a callable function from the library after it has been created from IR?

In this case I think Erick's suggestion is reasonable, we can define an MLIR attribute / LLVM metadata that indicates the preferred entry point.

> After some checks I think that we probably need to allow users to specify a complete signature of a top-level function. What do you think?
> Some details: I see an assertion failure [here](https://github.com/PennyLaneAI/catalyst/blob/1abfbc69c484ba8f4c6310b2eaa3e36ed0ec9748/frontend/catalyst/compilation_pipelines.py#L656). The failing test checks [compilation from IR string](https://github.com/PennyLaneAI/catalyst/blob/1abfbc69c484ba8f4c6310b2eaa3e36ed0ec9748/frontend/test/pytest/test_compiler.py#L297). The shape there refers to a Pytree shape, and I think user should already know it at the time they write the IR string.

Given the annotation, we should be able to read the type signature from the IR, at least at the MLIR level (like in your linked example). The LLVMIR case is more complicated though 🤔 

Now with the added Pytree support, the issue is complicated further. But I think we could change it so that in the absence of a PyTree shape specification (but given the type signature!), we can assume a flat tuple for args / results.

> Btw, do you think that the name shape better associates with tensor shapes rather than with the Pytree shapes? How about renaming it to pytree_shape or something like that? CC @maliasadi @erick-xanadu

Agree on this. JAX calls these `PyTreeDef`s or `treedef`s.

Reply:
I just changed the name of the preferred entry point to be "catalyst.entry_point". This is safer because no python functions can hold dots as their identifiers.